middle_end/flambda2/to_cmm/to_cmm_set_of_closures.ml

```diff
@@ -0,0 +1,431 @@
+  in
+  (* Init the env and create a jump id for the return continuation in case a
+     trap action is attached to one of its calls *)
+  let env = Env.enter_function_def env return_continuation exn_continuation in
+  (* Translate the arg list and body *)
+  let env, fun_args = C.bound_parameters env params in
+  let fun_body, res = translate_expr env res body in
+  let fun_flags =
+    if Flambda_features.optimize_for_speed () then [] else [Cmm.Reduce_code_size]
+  in
+  let linkage_name = Linkage_name.to_string (Code_id.linkage_name code_id) in
+  C.fundecl linkage_name fun_args fun_body fun_flags fun_dbg, res
+
+let params_and_body env res code_id p ~fun_dbg ~translate_expr =
+  Function_params_and_body.pattern_match p
+    ~f:(fun
+         ~return_continuation
+         ~exn_continuation
+         params
+         ~body
+         ~my_closure
+         ~is_my_closure_used
+         ~my_depth:_
+         ~free_names_of_body:_
+       ->
+      try
+        params_and_body0 env res code_id ~fun_dbg ~return_continuation
+          ~exn_continuation params ~body ~my_closure ~is_my_closure_used
+          ~translate_expr
+      with Misc.Fatal_error as e ->
+        Format.eprintf
+          "\n\
+           %sContext is:%s translating function %a to Cmm with return cont %a, \
+           exn cont %a and body:@ %a\n"
+          (Flambda_colours.error ())
+          (Flambda_colours.normal ())
+          Code_id.print code_id Continuation.print return_continuation
+          Continuation.print exn_continuation Expr.print body;
+        raise e)
+
+(* Translation of sets of closures. *)
+
+let debuginfo_for_set_of_closures env set =
+  let code_ids_in_set =
+    Set_of_closures.function_decls set
+    |> Function_declarations.funs |> Function_slot.Map.data
+  in
+  let dbg =
+    List.map
+      (fun code_id -> Env.get_code_metadata env code_id |> Code_metadata.dbg)
+      code_ids_in_set
+    |> List.sort Debuginfo.compare
+  in
+  (* Choose the debuginfo with the earliest source location. *)
+  match dbg with [] -> Debuginfo.none | dbg :: _ -> dbg
+
+let let_static_set_of_closures env symbs set (layout : Slot_offsets.layout)
+    ~prev_updates =
+  let set_of_closures_symbol_ref = ref None in
+  let fun_decls = Set_of_closures.function_decls set in
+  let decls = Function_declarations.funs fun_decls in
+  let value_slots = Set_of_closures.value_slots set in
+  let dbg = debuginfo_for_set_of_closures env set in
+  let l, length, env, _effs, updates =
+    Static.fill_layout ~set_of_closures_symbol_ref (Some symbs) decls dbg
+      ~startenv:layout.startenv value_slots env Ece.pure [] prev_updates
+      ~starting_offset:0 layout.slots
+  in
+  let block =
+    match l with
+    | _ :: _ ->
+      let header = C.cint (C.black_closure_header length) in
+      let sdef =
+        match !set_of_closures_symbol_ref with
+        | None -> []
+        | Some s ->
+          (* CR mshinwell: Is [global:false] correct? *)
```

Comment:
Yes it is: basically, the symbol for the whole set of closures is generated by `to_cmm` when some updates are generated (since those updates will update the field of the set using the symbol for the whole set as a reference point), so it is only ever used locally by the updates expressions, which should be local (i.e. in the same module/object file than the set of closures).

Reply:
OK